lsp/server/src/server.ts

```diff
@@ -32,32 +32,99 @@ const connection = createConnection(ProposedFeatures.all);
 
-class SA4UConfig {
-	compilationDir: string
-	constructor(compilationDir: string) {
-		this.compilationDir = '/src/' + compilationDir;
+class SA4UDebug {
+	Image: string;
+	MaintainContainerOnExit: boolean;
+	constructor(image:string, maintainOnExit:boolean) {
+		this.Image = image;
+		this.MaintainContainerOnExit = maintainOnExit;
 	}
 }
 
-class SA4UIgnoredFiles {
-	ignoreFilesCommand: string
-	constructor(ignoreFilesArray: string[]) {
-		if (ignoreFilesArray.length !== 0) {
-			this.ignoreFilesCommand = '-i '.concat(ignoreFilesArray.join(' -i '));
+class SA4UJSON {
+	ProtocolDefinitionFile: string;
+	CompilationDir: string;
+	PriorTypes: string;
+	IgnoreFiles?: string[];
+	Debug?:SA4UDebug;
+	constructor(config: any) {
+		this.ProtocolDefinitionFile = config.MessageDefinition || config.ProtocolDefinitionFile;
+		this.CompilationDir = config.CompilationDir;
+		this.PriorTypes = config.PriorTypes;
+		if (config.IgnoreFiles && config.IgnoreFiles !== '') {
+			this.IgnoreFiles = config.IgnoreFiles?.split?.('\n') || config.IgnoreFiles;
+		}
+		if (config.DebugMode === true) {
+			this.Debug = new SA4UDebug(config.DebugModeDockerImage, config.DebugModeMaintainContainerOnExit);
+		}
+	}
+	updateVSCodeConfig(): void {
+		connection.sendNotification('UpdateConfig', { param: 'SA4U.MessageDefinition', value: this.ProtocolDefinitionFile });
+		connection.sendNotification('UpdateConfig', { param: 'SA4U.CompilationDir', value: this.CompilationDir });
+		connection.sendNotification('UpdateConfig', { param: 'SA4U.PriorTypes', value: this.PriorTypes });
+		connection.sendNotification('UpdateConfig', { param: 'SA4U.IgnoreFiles', value: this.IgnoreFiles?.join('\n') || '' });
+		if (this.Debug) {
+			connection.sendNotification('UpdateConfig', { param: 'SA4U.DebugMode', value: true });
+			connection.sendNotification('UpdateConfig', { param: 'SA4U.DebugModeDockerImage', value: this.Debug.Image });
+			connection.sendNotification('UpdateConfig', { param: 'SA4U.DebugModeMaintainContainerOnExit', value: this.Debug.MaintainContainerOnExit });
+		}
+	}
+	async writeToJSON(): Promise<void> {
+		try {
+			await promises.writeFile(CONFIG_FILE, JSON.stringify(this, null, '    '));
+		} catch (err) {
+			console.warn(`Failed to write to ${CONFIG_FILE}: ${err}`);
+		}
+	}
+}
+
+class SA4UConfig {
+	// Required
+	messDef: string;
+	compilationDir: string;
+	priorTypes: string;
+	//Optional
+	ignoreFiles: string[];
+	image: string;
+	debugMaintainOnExit: boolean;
+	serializePath: string;
+	constructor(configJSON: any) {
+		if (configJSON !== '') {
+			this.messDef = configJSON['ProtocolDefinitionFile'];
+			this.compilationDir = configJSON['CompilationDir'];
+			this.priorTypes = configJSON['PriorTypes'];
+			this.ignoreFiles = configJSON['IgnoreFiles'];
+			this.image = configJSON['Debug']?.['Image'] || dockerImageName;
+			this.debugMaintainOnExit = configJSON['Debug']?.['MaintainContainerOnExit'] || false;
+			this.serializePath = '/src/.sa4u';
```

Comment:
We should should probably set `this.serializePath = '/src/.sa4u/cache'` (or something similar?) The SA4U configuration file lives in `.sa4u`, and there will be 100s of files written to the cache. We should try to keep its directory structure clean. *I think* that the SA4U Python tool creates the cache directory now, if it doesn't already exist.

Reply:
Done